no method named `property` found for struct `utoipa::openapi::AllOfBuilder` in the current scope

Hello,

While implementing ToSchema on a project, I got the following error:
```
error[E0599]: no method named `property` found for struct `utoipa::openapi::AllOfBuilder` in the current scope
  --> src/main.rs:14:21
   |
14 | #[derive(Serialize, ToSchema)]
   |                     ^^^^^^^^ method not found in `AllOfBuilder`
   |
   = note: this error originates in the derive macro `ToSchema` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I made a small test project to reproduce the issue:
```rust
use serde::Serialize;
use utoipa::ToSchema;

#[derive(Serialize, ToSchema)]
pub struct F {}

#[derive(Serialize, ToSchema)]
#[serde(tag = "test_c")]
pub enum C {
    D,
    E(F)
}

#[derive(Serialize, ToSchema)] // <-- Error happens here
#[serde(tag = "test_a")]
pub enum A {
    B {
        #[serde(flatten)]
        c: C
    }
}

fn main() {}
```

Here is the code generated with cargo expand:

<details>

<summary>Cargo expand</summary>

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use serde::Serialize;
use utoipa::ToSchema;
pub struct F {}
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl _serde::Serialize for F {
        fn serialize<__S>(
            &self,
            __serializer: __S,
        ) -> _serde::__private::Result<__S::Ok, __S::Error>
        where
            __S: _serde::Serializer,
        {
            let __serde_state = _serde::Serializer::serialize_struct(
                __serializer,
                "F",
                false as usize,
            )?;
            _serde::ser::SerializeStruct::end(__serde_state)
        }
    }
};
impl utoipa::__dev::ComposeSchema for F {
    fn compose(
        mut generics: Vec<utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>>,
    ) -> utoipa::openapi::RefOr<utoipa::openapi::schema::Schema> {
        {
            let mut object = utoipa::openapi::ObjectBuilder::new();
            object
        }
            .into()
    }
}
impl utoipa::ToSchema for F {
    fn name() -> std::borrow::Cow<'static, str> {
        std::borrow::Cow::Borrowed("F")
    }
    fn schemas(
        schemas: &mut Vec<
            (String, utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>),
        >,
    ) {
        schemas.extend([]);
    }
}
#[serde(tag = "test_c")]
pub enum C {
    D,
    E(F),
}
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl _serde::Serialize for C {
        fn serialize<__S>(
            &self,
            __serializer: __S,
        ) -> _serde::__private::Result<__S::Ok, __S::Error>
        where
            __S: _serde::Serializer,
        {
            match *self {
                C::D => {
                    let mut __struct = _serde::Serializer::serialize_struct(
                        __serializer,
                        "C",
                        1,
                    )?;
                    _serde::ser::SerializeStruct::serialize_field(
                        &mut __struct,
                        "test_c",
                        "D",
                    )?;
                    _serde::ser::SerializeStruct::end(__struct)
                }
                C::E(ref __field0) => {
                    _serde::__private::ser::serialize_tagged_newtype(
                        __serializer,
                        "C",
                        "E",
                        "test_c",
                        "E",
                        __field0,
                    )
                }
            }
        }
    }
};
impl utoipa::__dev::ComposeSchema for C {
    fn compose(
        mut generics: Vec<utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>>,
    ) -> utoipa::openapi::RefOr<utoipa::openapi::schema::Schema> {
        Into::<
            utoipa::openapi::schema::OneOfBuilder,
        >::into(utoipa::openapi::OneOf::with_capacity(2usize))
            .item(
                utoipa::openapi::schema::Object::builder()
                    .property(
                        "test_c",
                        utoipa::openapi::schema::Object::builder()
                            .schema_type(
                                utoipa::openapi::schema::SchemaType::new(
                                    utoipa::openapi::schema::Type::String,
                                ),
                            )
                            .enum_values::<[&str; 1usize], &str>(Some(["D"])),
                    )
                    .required("test_c"),
            )
            .item(
                utoipa::openapi::schema::AllOfBuilder::new()
                    .item(
                        utoipa::openapi::schema::RefBuilder::new()
                            .ref_location_from_schema_name({
                                let res = ::alloc::fmt::format(
                                    format_args!("{0}", <F as utoipa::ToSchema>::name()),
                                );
                                res
                            }),
                    )
                    .item(
                        utoipa::openapi::schema::Object::builder()
                            .property(
                                "test_c",
                                utoipa::openapi::schema::Object::builder()
                                    .schema_type(
                                        utoipa::openapi::schema::SchemaType::new(
                                            utoipa::openapi::schema::Type::String,
                                        ),
                                    )
                                    .enum_values::<[&str; 1usize], &str>(Some(["E"])),
                            )
                            .required("test_c"),
                    ),
            )
            .into()
    }
}
impl utoipa::ToSchema for C {
    fn name() -> std::borrow::Cow<'static, str> {
        std::borrow::Cow::Borrowed("C")
    }
    fn schemas(
        schemas: &mut Vec<
            (String, utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>),
        >,
    ) {
        schemas
            .extend([
                (
                    String::from({
                        let res = ::alloc::fmt::format(
                            format_args!("{0}", <F as utoipa::ToSchema>::name()),
                        );
                        res
                    }),
                    <F as utoipa::PartialSchema>::schema(),
                ),
            ]);
        <F as utoipa::ToSchema>::schemas(schemas);
    }
}
#[serde(tag = "test_a")]
pub enum A {
    B { #[serde(flatten)] c: C },
}
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl _serde::Serialize for A {
        fn serialize<__S>(
            &self,
            __serializer: __S,
        ) -> _serde::__private::Result<__S::Ok, __S::Error>
        where
            __S: _serde::Serializer,
        {
            match *self {
                A::B { ref c } => {
                    let mut __serde_state = _serde::Serializer::serialize_map(
                        __serializer,
                        _serde::__private::None,
                    )?;
                    _serde::ser::SerializeMap::serialize_entry(
                        &mut __serde_state,
                        "test_a",
                        "B",
                    )?;
                    _serde::Serialize::serialize(
                        &c,
                        _serde::__private::ser::FlatMapSerializer(&mut __serde_state),
                    )?;
                    _serde::ser::SerializeMap::end(__serde_state)
                }
            }
        }
    }
};
impl utoipa::__dev::ComposeSchema for A {
    fn compose(
        mut generics: Vec<utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>>,
    ) -> utoipa::openapi::RefOr<utoipa::openapi::schema::Schema> {
        Into::<
            utoipa::openapi::schema::OneOfBuilder,
        >::into(utoipa::openapi::OneOf::with_capacity(1usize))
            .item(
                utoipa::openapi::AllOfBuilder::new()
                    .item(
                        utoipa::openapi::schema::RefBuilder::new()
                            .ref_location_from_schema_name({
                                let res = ::alloc::fmt::format(
                                    format_args!("{0}", <C as utoipa::ToSchema>::name()),
                                );
                                res
                            }),
                    )
                    .item({
                        let mut object = utoipa::openapi::ObjectBuilder::new();
                        object
                    })
                    .property(
                        "test_a",
                        utoipa::openapi::schema::Object::builder()
                            .schema_type(
                                utoipa::openapi::schema::SchemaType::new(
                                    utoipa::openapi::schema::Type::String,
                                ),
                            )
                            .enum_values::<[&str; 1usize], &str>(Some(["B"])),
                    )
                    .required("test_a"),
            )
            .into()
    }
}
impl utoipa::ToSchema for A {
    fn name() -> std::borrow::Cow<'static, str> {
        std::borrow::Cow::Borrowed("A")
    }
    fn schemas(
        schemas: &mut Vec<
            (String, utoipa::openapi::RefOr<utoipa::openapi::schema::Schema>),
        >,
    ) {
        schemas
            .extend([
                (
                    String::from({
                        let res = ::alloc::fmt::format(
                            format_args!("{0}", <C as utoipa::ToSchema>::name()),
                        );
                        res
                    }),
                    <C as utoipa::PartialSchema>::schema(),
                ),
            ]);
        <C as utoipa::ToSchema>::schemas(schemas);
    }
}
fn main() {}
```

</details>

--- a/utoipa-gen/CHANGELOG.md
+++ b/utoipa-gen/CHANGELOG.md
@@ -1,5 +1,11 @@
 # Changelog - utoipa-gen
 
+## Unreleased
+
+### Fixed
+
+* Fix tagged enum with flatten fields (https://github.com/juhaku/utoipa/pull/1208)
+
 ## 5.2.0 - Nov 2024
 
 ### Fixed
@@ -454,7 +460,7 @@
 * Add support for multiple examples in response (https://github.com/juhaku/utoipa/pull/403)
 * Add Example type to OpenApi types (https://github.com/juhaku/utoipa/pull/402)
 * Add derive info support for derive OpenApi (https://github.com/juhaku/utoipa/pull/400)
-* Add `merge` functionality for `OpenApi` (https://github.com/juhaku/utoipa/pull/397) 
+* Add `merge` functionality for `OpenApi` (https://github.com/juhaku/utoipa/pull/397)
 * Add derive servers attribute for OpenApi (https://github.com/juhaku/utoipa/pull/395)
 * Add support for unit sructs (https://github.com/juhaku/utoipa/pull/392)
 * Add support for `schema_with` custom fn reference (https://github.com/juhaku/utoipa/pull/390)

--- a/utoipa-gen/src/component/schema.rs
+++ b/utoipa-gen/src/component/schema.rs
@@ -313,6 +313,7 @@ pub struct NamedStructSchema {
     pub schema_as: Option<As>,
     fields_references: Vec<SchemaReference>,
     bound: Option<Bound>,
+    is_all_of: bool,
 }
 
 #[cfg_attr(feature = "debug", derive(Debug))]
@@ -385,6 +386,7 @@ impl NamedStructSchema {
             .flatten()
             .collect::<Vec<_>>();
 
+        let mut object_tokens_empty = true;
         let object_tokens = fields_vec
             .iter()
             .filter(|(_, field_rules, ..)| !field_rules.skip && !field_rules.flatten)
@@ -415,6 +417,7 @@ impl NamedStructSchema {
                     _field,
                     field_schema,
                 )| {
+                    object_tokens_empty = false;
                     let rename_to = field_rules
                         .rename
                         .as_deref()
@@ -513,8 +516,13 @@ impl NamedStructSchema {
                 tokens.extend(quote! {
                     utoipa::openapi::AllOfBuilder::new()
                         #flattened_tokens
-                    .item(#object_tokens)
+
                 });
+                if !object_tokens_empty {
+                    tokens.extend(quote! {
+                        .item(#object_tokens)
+                    });
+                }
                 true
             }
         } else {
@@ -552,6 +560,7 @@ impl NamedStructSchema {
             schema_as,
             fields_references,
             bound,
+            is_all_of: all_of,
         })
     }
 

--- a/utoipa-gen/src/component/schema/enums.rs
+++ b/utoipa-gen/src/component/schema/enums.rs
@@ -520,11 +520,13 @@ impl MixedEnumContent {
                 let mut schema_tokens = schema.to_token_stream();
                 (
                     if schema.is_all_of {
-                        let object_builder_tokens = quote! { utoipa::openapi::schema::Object::builder() };
-                        let enum_schema_tokens = EnumSchema::<ObjectSchema>::tagged(object_builder_tokens)
-                            .tag(tag, PlainSchema::for_name(name.as_ref()))
-                            .features(enum_features)
-                            .to_token_stream();
+                        let object_builder_tokens =
+                            quote! { utoipa::openapi::schema::Object::builder() };
+                        let enum_schema_tokens =
+                            EnumSchema::<ObjectSchema>::tagged(object_builder_tokens)
+                                .tag(tag, PlainSchema::for_name(name.as_ref()))
+                                .features(enum_features)
+                                .to_token_stream();
                         schema_tokens.extend(quote! {
                             .item(#enum_schema_tokens)
                         });

--- a/utoipa-gen/src/component/schema/enums.rs
+++ b/utoipa-gen/src/component/schema/enums.rs
@@ -517,12 +517,24 @@ impl MixedEnumContent {
                     MixedEnumContent::split_enum_features(variant_features);
                 let schema = NamedStructSchema::new(root, fields, variant_features)?;
 
-                let schema_tokens = schema.to_token_stream();
+                let mut schema_tokens = schema.to_token_stream();
                 (
-                    EnumSchema::<ObjectSchema>::tagged(schema_tokens)
-                        .tag(tag, PlainSchema::for_name(name.as_ref()))
-                        .features(enum_features)
-                        .to_token_stream(),
+                    if schema.is_all_of {
+                        let object_builder_tokens = quote! { utoipa::openapi::schema::Object::builder() };
+                        let enum_schema_tokens = EnumSchema::<ObjectSchema>::tagged(object_builder_tokens)
+                            .tag(tag, PlainSchema::for_name(name.as_ref()))
+                            .features(enum_features)
+                            .to_token_stream();
+                        schema_tokens.extend(quote! {
+                            .item(#enum_schema_tokens)
+                        });
+                        schema_tokens
+                    } else {
+                        EnumSchema::<ObjectSchema>::tagged(schema_tokens)
+                            .tag(tag, PlainSchema::for_name(name.as_ref()))
+                            .features(enum_features)
+                            .to_token_stream()
+                    },
                     schema.fields_references,
                 )
             }

--- a/utoipa-gen/tests/schema_derive_test.rs
+++ b/utoipa-gen/tests/schema_derive_test.rs
@@ -1090,6 +1090,55 @@ fn derive_simple_enum_serde_tag() {
 }
 
 #[test]
+fn derive_simple_enum_serde_tag_with_flatten_content() {
+    #[derive(Serialize, ToSchema)]
+    #[allow(unused)]
+    struct Foo {
+        name: &'static str,
+    }
+
+    let value: Value = api_doc! {
+        #[derive(Serialize)]
+        #[serde(tag = "tag")]
+        enum Bar {
+            One {
+                #[serde(flatten)]
+                foo: Foo,
+            },
+        }
+    };
+
+    assert_json_eq!(
+        value,
+        json!({
+            "oneOf": [
+                {
+                    "allOf": [
+                        {
+                            "$ref": "#/components/schemas/Foo",
+                        },
+                         {
+                            "type": "object",
+                            "properties":  {
+                                "tag":  {
+                                    "type": "string",
+                                    "enum":  [
+                                        "One",
+                                    ],
+                                },
+                            },
+                            "required":  [
+                                "tag",
+                            ],
+                        },
+                    ],
+                },
+            ],
+        })
+    );
+}
+
+#[test]
 fn derive_simple_enum_serde_untagged() {
     let value: Value = api_doc! {
         #[derive(Serialize)]

